framework/platform/bpl/uci/db/bpl_db.cpp

```diff
@@ -17,45 +18,94 @@
 namespace beerocks {
 namespace bpl {
 
+#define DB_FILE "prplMesh_db"
+
 bool db_has_entry(const std::string &entry_type, const std::string &entry_name)
 {
-    LOG(TRACE) << " - NOT IMPLEMENTED!";
-    return false;
+    LOG(TRACE) << entry_type << ":" << entry_name;
+    return uci_entry_exists(DB_FILE, entry_type, entry_name);
 }
 
 bool db_add_entry(const std::string &entry_type, const std::string &entry_name,
                   const std::unordered_map<std::string, std::string> &params)
 {
-    LOG(TRACE) << " - NOT IMPLEMENTED!";
-    return true;
+    LOG(TRACE) << entry_type << ":" << entry_name;
+
+    // Check if entry of the same name (and any type) exist.
+    if (!uci_entry_exists(DB_FILE, "", entry_name)) {
+        LOG(DEBUG) << "Entry " << entry_name << " does not exist! Create new one.";
+        if (!uci_add_entry(DB_FILE, entry_type, entry_name)) {
+            LOG(ERROR) << "Failed to create entry " << entry_name << "!";
+            return false;
+        }
+    } else {
+        // Entry exists, validate type match
+        if (uci_entry_exists(DB_FILE, entry_type, entry_name)) {
+            LOG(ERROR) << "Entry " << entry_name << " exists, but under diffrent type!";
+            return false;
+        }
+    }
+    // Update the new/existing entry
+    LOG(DEBUG) << "Update entry " << entry_name << " with new values.";
+    return uci_set_entry(DB_FILE, entry_type, entry_name, params);
 }
 
 bool db_set_entry(const std::string &entry_type, const std::string &entry_name,
                   const std::unordered_map<std::string, std::string> &params)
 {
-    LOG(TRACE) << " - NOT IMPLEMENTED!";
-    return true;
+    LOG(TRACE) << entry_type << ":" << entry_name;
+
+    if (!uci_entry_exists(DB_FILE, entry_type, entry_name)) {
+        LOG(ERROR) << "Entry " << entry_name << " not found, or exists under a diffrent type!";
+        return false;
+    }
+    // Update the new/existing entry
+    LOG(DEBUG) << "Update entry " << entry_name << " with new values.";
+    return uci_set_entry(DB_FILE, entry_type, entry_name, params);
 }
 
 bool db_get_entry(const std::string &entry_type, const std::string &entry_name,
                   std::unordered_map<std::string, std::string> &params)
 {
-    LOG(TRACE) << " - NOT IMPLEMENTED!";
-    return true;
+    LOG(TRACE) << entry_type << ":" << entry_name;
+
+    return uci_get_entry(DB_FILE, entry_type, entry_name, params);
 }
 
 bool db_get_entries_by_type(
     const std::string &entry_type,
```

Comment:
both entry name and type are strings is this on purpose?

Reply:
Yes, I don't want to force the user to use a preset amount of types.
This way each place that uses the UCI persistent DB can use their own types